pyccel/ast/core.py

```diff
@@ -2022,7 +2025,7 @@ def __new__(
         if isinstance(dtype, str) or str(dtype) == '*':
 
             dtype = datatype(str(dtype))
-        elif not isinstance(dtype, DataType):
+        elif not isinstance(dtype, (DataType, NativeTuple)):
             raise TypeError('datatype must be an instance of DataType.')
```

Comment:
This change is something I pushed a few hours ago. I was originally subclassing DataType, but the problem is that DataType subclasses Singleton. Singleton ensures that only one instance of this class exists. This makes sense for objects such as NativeInt which have a fixed type however it is useful for the tuple to include information about it's contents. In order for this to work each different tuple needs its own NativeTuple which then contains NativeInts etc.

It may be possible to modify the usage so that the subclass types are determined from the variable. I will try to modify this

Reply:
I see the problem now. Theoretically speaking, each call to `tuple()` defines a new DataType, which depends on the type of the arguments. Sometimes we only care if an object is a NativeTuple, and sometimes we really want to know its specific type.